packages/apps/tools/src/components/Global/AccountNameField/index.tsx

```diff
@@ -1,41 +1,135 @@
-import { ITextFieldProps, SystemIcon, TextField } from '@kadena/react-ui';
+import {
+  IconButton,
+  IInputProps,
+  IInputWrapperProps,
+  Input,
+  InputWrapper,
+  ISelectProps,
+  Option,
+  Select,
+  SystemIcon,
+} from '@kadena/react-ui';
 
+import { accountInputWrapperStyle } from './styles.css';
+
+import { useWalletConnectClient } from '@/context/connect-wallet-context';
+import { useDidUpdateEffect } from '@/hooks';
 import useTranslation from 'next-translate/useTranslation';
-import React, { type FC } from 'react';
+import React, { ChangeEvent, FC, useState } from 'react';
 import { FieldError } from 'react-hook-form';
 import * as z from 'zod';
 
 interface IAccountNameFieldProps
-  extends Partial<Omit<ITextFieldProps, 'inputProps'>> {
-  inputProps: Partial<ITextFieldProps['inputProps']>;
+  extends Partial<Omit<IInputWrapperProps, 'children'>> {
+  inputProps: Partial<
+    Pick<
+      IInputProps,
+      'id' | 'placeholder' | 'ref' | 'name' | 'onChange' | 'onBlur'
+    >
+  >;
   error?: FieldError;
 }
 
-// @see; https://github.com/kadena-io/chainweb-node/blob/master/pact/coin-contract/v5/coin-v5.pact#L14
 export const NAME_VALIDATION = z.string().trim().min(3).max(256);
 
-const AccountNameField: FC<IAccountNameFieldProps> = ({
-  error,
+export const AccountNameField: FC<IAccountNameFieldProps> = ({
   inputProps,
   status,
   helperText,
+  error,
   ...rest
 }) => {
+  const elementId = 'kd-select-account-input';
+  const [mode, setMode] = useState<'input' | 'select'>('input');
   const { t } = useTranslation('common');
+  const { onChange, ...restInputProps } = inputProps;
+
+  const {
+    accounts,
+    selectedAccount,
+    selectedNetwork,
+    setSelectedAccount,
+    session,
+  } = useWalletConnectClient();
+
+  useDidUpdateEffect(() => {
+    if (session && accounts?.length) setMode('select');
+    else setMode('input');
+  }, [session, accounts]);
+
+  const getAccounts = (): string[] => {
+    // eslint-disable-next-line @rushstack/security/no-unsafe-regexp
+    const regex = new RegExp(`kadena\:${selectedNetwork}.*`);
+    const result = new Set(
+      accounts
+        ?.filter((account) => {
+          if (account.match(regex)) return account;
+        })
+        ?.map((account) => {
+          const [, , accountName] = account.split(':');
+          return accountName;
+        }),
+    );
+
+    return Array.from(result) ?? [];
+  };
```

Comment:
- you can use `"".includes()` (no need to use regex for that, esp when you have to ignore a lint rule)
- you can return boolean in a filter lambda (no need to re-return when you're in a filter)
- you can pick the 3rd element straight from the split action
- A `Array.from(<SET>)` always returns an array, even if the set is empty

```suggestion
  const getAccounts = (): string[] => {
    return Array.from(new Set(
      accounts
        ?.filter((account) => account.includes(`kadena\:${selectedNetwork}`))
        ?.map((account) => account.split(':')[3]),
    ));
  };
```

Reply:
valid point, all fixed